sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -172,13 +188,23 @@ object NestedColumnAliasing {
           (f, Alias(f, s"_gen_alias_${exprId.id}")(exprId, Seq.empty, None))
         }
 
+
+        // Do deduplication based on semanticEquals, and then sum.
+        val nestedFieldNum = nestedFieldToAlias
+          .foldLeft(Seq[ExtractValue]()) {
+            (unique, curr) => if (!unique.exists(curr._1.semanticEquals(_))) {
+              curr._1 +: unique
+            } else {
+              unique
+            }
+          }
+          .map { t => totalFieldNum(t.dataType)  }
+          .sum
```

Comment:
Could you check this part, @viirya ? also cc: @dongjoon-hyun because IIUC he is an orignal author of this file: https://github.com/apache/spark/commit/257391497bea51553d5a9e420c20a3fdfe4d36a9#diff-43334bab9616cc53e8797b9afa9fc7aaR117-R126

Reply:
Why we need to deduplicate again?

Reply:
Take a look at this query in comment below: https://github.com/apache/spark/pull/28898#discussion_r448718366

If we don't do deduplication, then the calculation will gives us `3`. Then by failing this if-statement, it assumes all the nested columns in `name` is referenced, which isn't the case. Thus, deduplication is needed.

Reply:
Like:

```scala

def removeQualifier(f: ExtractValue): ExtractValue = {
  f.transform {
    case a: AttributeReference => a.withQualifier(Seq.empty)
  }.asInstanceOf[ExtractValue]
}
```

```scala
val dedupNestedFields = nestedFields.filter {
  case e @ (_: GetStructField | _: GetArrayStructFields) =>
    val child = e.children.head
    nestedFields.forall(f => child.find(_.semanticEquals(f)).isEmpty)
  case _ => true
}.map(removeQualifier)
```

And when we need to query the map, we do:
```scala
nestedFieldToAlias.contains(removeQualifier(f))
```

or 

```scala
nestedFieldToAlias(removeQualifier(f)).toAttribute
```

Reply:
The problem is your approach is, which one is selected from the `ExtractValue`s with different qualifier, is non-deterministic. 

Later when we query the `nestedFieldToAlias` map, you might fail to find the corresponding item from the map due to qualifier difference.

I think the safer approach is to clean up all qualifier.

Reply:
Hmm, I don't quite follow. All I change is to not double-counting duplicated nested columns. The counting method doesn't even care about the expression itself.

Reply:
I think you might reply in the wrong comments, but for that `groupBy`, only `exprId` and `dataType` is referenced. So the situation you described shouldn't matter.

Reply:
No, I mean this comment thread.

I am not sure if you are aware of it. The reason you need to deduplicate here, is because the semantically same `ExtractValue`s apply on attributes with different qualifier, e.g. there are two `name.first`, but one refers to `name` with qualifier `a` and another refers to qualifier `b`.

I did a test using your query and cleaned up all qualifiers, it works well.

And what I said in above comment is, you select arbitrary one `ExtractValue` from these `ExtractValue` with different qualifiers, but later we will look into the map using given `ExtractValue`. You might fail a case that you select the `name.first` with qualifier `a`, but later you look at the map using `name.first` with qualifier `b`.

Reply:
Ah, sorry, I thought you deduplicate when producing `nestedFieldToAlias`. But you don't touch it.

Reply:
No, I mean you probably misread the code here. This map still returns `ExprId->Seq(name.last, name.first, name.first)` because I didn't change the lookup map here. All I change is to count the field number differently so that it won't trigger the `else` statement -- which means all leaf nodes are covered, no schema pruning is required.

Reply:
Yeah, I didn't touch `nestedFieldToAlias`.

Reply:
I see. I can see the bad thing about this is, you will have duplicate aliases for the semantically same `ExtractValue`s, e.g. two aliases for two `name.first`s.

It may not be a big deal here, but it is possibly in the optimized query plan, you will have multiple gen_alias_xxx which refer to the same `ExtractValue`.

Reply:
I agree, but I think this is outside this PR's scope that I first want to iterate this thing safely.